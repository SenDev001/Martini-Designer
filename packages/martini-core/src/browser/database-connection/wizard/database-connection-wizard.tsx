import { SelectionService } from "@theia/core";
import { OpenerService } from "@theia/core/lib/browser";
import URI from "@theia/core/lib/common/uri";
import { Field, Form, Formik, FormikErrors, FormikProps } from "formik";
import { inject, injectable } from "inversify";
import { noop } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import * as Yup from "yup";
import {
    DatabaseType,
    getTypeLabel, MartiniDatabaseConnectionManager
} from "../../../common/database-connection/martini-database-connection-manager";
import { UpDownLabel } from "../../components/up-down-label";
import { FormEffect, FormRow, OnFormChange, validateSchema } from "../../form/form";
import { Progress } from "../../progress/progress-service";
import { AbstractWizard, AbstractWizardPage, Wizard } from "../../wizard/wizard";
import { WizardContribution } from "../../wizard/wizard-contribution";
import { isDbConnectionSelection } from "../database-connection-contribution";
import { DatabaseConnectionEditor } from "../editor/database-connection-editor";

@injectable()
export class DatabaseConnectionWizardContribution implements WizardContribution {
    @inject(SelectionService)
    private selectionService: SelectionService;
    @inject("Factory<DatabaseConnectionWizard>")
    private factory: () => DatabaseConnectionWizard;

    readonly wizardType = "new";
    readonly id = "database-connection";
    readonly label = messages.database_connection;
    readonly description = messages.create_db_connection;
    readonly iconClass = "martini-icon martini-database-icon";
    readonly keybinding = "ctrlcmd+alt+n d";
    readonly primary = true;

    async createWizard(): Promise<Wizard> {
        return this.factory();
    }

    isVisible() {
        const selection = this.selectionService.selection;
        return isDbConnectionSelection(selection) || (selection instanceof Array && selection[0] === "Databases");
    }

}

@injectable()
export class DatabaseConnectionWizardPage extends AbstractWizardPage {
    existingNames: string[] = [];
    private readonly defaultConfig: DatabaseConnectionConfig = {
        type: DatabaseType.JDBC,
        name: "MyConnection"
    };
    config: DatabaseConnectionConfig = this.defaultConfig;

    render(): React.ReactNode {
        return <CreateDatabaseConnectionForm
            existingNames={this.existingNames}
            defaultConfig={this.defaultConfig}
            onValidate={errors => this.handleValidate(errors)}
            onChange={config => this.handleChange(config)}
        />;
    }

    private handleValidate(errors: FormikErrors<DatabaseConnectionConfig>) {
        this.complete = Object.keys(errors).length === 0;
    }

    private handleChange(config: DatabaseConnectionConfig) {
        this.config = config;
    }
}

export interface DatabaseConnectionConfig {
    type: DatabaseType;
    name: string;
}

export interface CreateDatabaseConnectionFormProps {
    existingNames: string[];
    defaultConfig?: DatabaseConnectionConfig;
    onValidate: (errors: FormikErrors<DatabaseConnectionConfig>) => void;
    onChange: (config: DatabaseConnectionConfig) => void;
}

const databaseConnectionNameRegExp = /^[a-zA-Z0-9\\_-]+$/;

export const CreateDatabaseConnectionForm: React.FC<CreateDatabaseConnectionFormProps> = (
    { existingNames, defaultConfig, onValidate, onChange }) => {

    const schema = React.useMemo(() => Yup.object().shape<Omit<DatabaseConnectionConfig, "type">>({
        name: Yup.string()
            .required()
            .max(30)
            .matches(databaseConnectionNameRegExp, messages.only_alphanumeric)
            .test("unique-name", messages.connection_name_exists, value => !existingNames.includes(value!))
    }), []);
    const validate = (config: DatabaseConnectionConfig) => validateSchema(schema, config, onValidate);

    const handleKeyDown = (e: React.KeyboardEvent, props: FormikProps<DatabaseConnectionConfig>) => {
        if (e.key !== "ArrowUp" && e.key !== "ArrowDown")
            return;
        const types = Object.values(DatabaseType);
        let index = types.indexOf(props.values.type);
        if (e.key === "ArrowUp")
            index = index - 1 < 0 ? types.length - 1 : index - 1;
        else if (e.key === "ArrowDown")
            index = index + 1 >= types.length ? 0 : index + 1;
        const type = types[index];
        props.setValues({
            ...props.values,
            type
        });
        e.preventDefault();
    };

    return <Formik
        initialValues={defaultConfig}
        initialTouched={{
            type: true,
            name: true
        }}
        validate={validate}
        validateOnMount={true}
        onSubmit={noop}
    >
        {props => (
            <Form style={{
                display: "grid",
                gridColumnGap: 10,
                gridTemplateColumns: "max-content 1fr"
            }}>
                <FormEffect deps={[existingNames]} />
                <OnFormChange onChange={onChange} />
                <FormRow name="type" label={messages.type}>
                    <Field as="select" name="type" className="theia-select">
                        {Object.values(DatabaseType).map(type => (
                            <option key={type}
                                value={type}>{getTypeLabel(type)}</option>))}
                    </Field>
                </FormRow>
                <FormRow name="name" label={messages.name} gridTemplateColumns="1fr max-content">
                    <Field
                        type="text"
                        name="name"
                        className="theia-input"
                        autoFocus
                        onKeyDown={(e: React.KeyboardEvent) => handleKeyDown(e, props)}
                    />
                    <UpDownLabel toolTip={messages.pressing_updown_changes_type} />
                </FormRow>
            </Form>)}
    </Formik>;
};

@injectable()
export class DatabaseConnectionWizard extends AbstractWizard {
    readonly title = messages.create_db_connection_title;

    constructor(
        @inject(MartiniDatabaseConnectionManager)
        private readonly dbConnectionMngr: MartiniDatabaseConnectionManager,
        @inject(DatabaseConnectionWizardPage)
        private readonly page: DatabaseConnectionWizardPage,
        @inject(OpenerService)
        private readonly openerService: OpenerService) {
        super();
        this.pages.push(page);
    }

    async init(progress: Progress): Promise<void> {
        progress.report({
            message: messages.fetching_database_connections
        });
        const connections = await this.dbConnectionMngr.getAll();
        this.page.existingNames = connections.map(c => c.name);
        return super.init(progress);
    }

    async performFinish(): Promise<boolean> {
        const uri: URI = new URI(`${DatabaseConnectionEditor.URI_SCHEME}://${encodeURI(this.page.config.name)}`
            + `?type=${this.page.config.type}&new=true`);
        const handler = await this.openerService.getOpener(uri);
        handler.open(uri);
        return true;
    }

}
