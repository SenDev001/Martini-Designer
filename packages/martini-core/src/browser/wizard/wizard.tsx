import { Disposable, DisposableCollection, Emitter, Event } from "@theia/core";
import { injectable } from "inversify";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import { BaseDialog } from "../dialogs/dialogs";
import { Progress } from "../progress/progress-service";

export interface Wizard extends Disposable {
    currentPage: WizardPage | undefined;
    firstPage: WizardPage;
    title: string;
    onCompleteChanged: Event<void>;
    onUpdate: Event<void>;
    hasNext: boolean;
    hasPrevious: boolean;
    initialSize?: { width?: number, height?: number; };

    init(progress: Progress): Promise<void>;

    performFinish(progress: Progress): Promise<boolean>;

    canFinish(): boolean;

    getNextPage(current: WizardPage | undefined): WizardPage | undefined;

    getPreviousPage(current: WizardPage | undefined): WizardPage | undefined;

    render(): React.ReactNode;

    next(): void;

    previous(): void;
}

export interface WizardPage extends Disposable {
    description?: string;
    complete: boolean;
    onCompleteChanged: Event<void>;
    onUpdate: Event<void>;

    render(): React.ReactNode;

    dispose(): void;
}

@injectable()
export abstract class AbstractWizard implements Wizard {
    protected onCompleteChangedEmitter = new Emitter<void>();
    readonly onCompleteChanged = this.onCompleteChangedEmitter.event;
    protected onUpdateEmitter = new Emitter<void>();
    readonly onUpdate = this.onUpdateEmitter.event;

    abstract title: string;
    protected pages: WizardPage[] = [];
    protected history: WizardPage[] = [];
    protected current: WizardPage;
    protected toDispose = new DisposableCollection();

    /**
     * Initializes the wizard. If this method is overridden it should call super when all the pages have been added already.
     *
     * @param progress used to report progress
     */
    async init(progress: Progress): Promise<void> {
        if (this.pages.length) this.current = this.firstPage;
        this.toDispose.pushAll(this.pages);
    }

    abstract performFinish(progress: Progress): Promise<boolean>;

    canFinish(): boolean {
        return (
            this.pages.length === 0 ||
            (this.currentPage !== undefined &&
                this.currentPage.complete &&
                this.getNextPage(this.current) === undefined)
        );
    }

    getNextPage(current: WizardPage | undefined): WizardPage | undefined {
        if (!current) return undefined;
        const index = this.pages.indexOf(current);
        if (index >= 0 && index + 1 < this.pages.length)
            return this.pages[index + 1];
        return undefined;
    }

    getPreviousPage(current: WizardPage | undefined): WizardPage | undefined {
        return this.history.length === 0 ? undefined : this.history[this.history.length - 1];
    }

    render(): React.ReactNode {
        const page = this.currentPage;
        if (page) {
            return (
                <div style={{ display: "grid", overflowY: "hidden" }}>
                    {page.description && (
                        <>
                            <div
                                style={{
                                    marginBottom: 5,
                                    padding: "6px 12px"
                                }}
                            >
                                {page.description}
                            </div>
                            <div
                                style={{
                                    borderTop: "1px solid var(--theia-panel-border)",
                                    width: "100%",
                                    marginBottom: 5
                                }}
                            />
                        </>
                    )}
                    <div style={{ padding: "6px 12px", overflowY: "auto" }}>{page.render()}</div>
                </div>
            );
        }
        return <></>;
    }

    get firstPage() {
        return this.pages[0];
    }

    get currentPage() {
        return this.current;
    }

    get hasNext() {
        return this.pages.length > 1;
    }

    get hasPrevious() {
        return this.hasNext;
    }

    next(): void {
        if (this.current && this.current.complete) {
            const next = this.getNextPage(this.current);
            if (next) {
                this.history.push(this.current);
                this.current = next;
                this.onCompleteChangedEmitter.fire();
                this.onUpdateEmitter.fire();
            }
        }
    }

    previous(): void {
        if (this.current) {
            const previous = this.getPreviousPage(this.current);
            if (previous) {
                this.current = previous;
                this.history.pop();
                this.onCompleteChangedEmitter.fire();
                this.onUpdateEmitter.fire();
            }
        }
    }

    dispose(): void {
        this.toDispose.dispose();
    }
}

@injectable()
export abstract class AbstractWizardPage implements WizardPage {
    protected _complete = false;
    protected onCompleteChangedEmitter = new Emitter<void>();
    readonly onCompleteChanged = this.onCompleteChangedEmitter.event;
    protected onUpdateEmitter = new Emitter<void>();
    readonly onUpdate = this.onUpdateEmitter.event;
    protected toDispose = new DisposableCollection();
    protected _description = "";

    abstract render(): React.ReactNode;

    get complete() {
        return this._complete;
    }

    set complete(complete: boolean) {
        this._complete = complete;
        this.onCompleteChangedEmitter.fire();
    }

    set description(description: string) {
        if (this._description !== description) {
            this._description = description;
            this.onUpdateEmitter.fire();
        }
    }

    get description() {
        return this._description;
    }

    dispose(): void {
        this.toDispose.dispose();
    }
}

export interface WizardDialogProps {
    acceptLabel?: string;
    width?: number;
    height?: number;
}

export class WizardDialog extends BaseDialog<boolean> {
    protected readonly previousBtn: HTMLButtonElement | undefined;
    protected readonly nextBtn: HTMLButtonElement | undefined;
    protected readonly currentPageDisposable = new DisposableCollection();
    protected oldCurrent: WizardPage | undefined;
    private performingFinish = false;
    value = false;

    constructor(protected readonly wizard: Wizard, props?: WizardDialogProps) {
        super({
            title: wizard.title
        });
        const acceptLabel =
            props && props.acceptLabel ? props.acceptLabel : messages.finish;

        if (wizard.hasPrevious) {
            this.previousBtn = this.appendButton(this.createButton(messages.previous));
            this.previousBtn.addEventListener("click", () => this.wizard.previous());
        }
        if (wizard.hasNext) {
            this.nextBtn = this.appendButton(this.createButton(messages.next));
            this.nextBtn.addEventListener("click", () => this.wizard.next());
        }
        this.appendCloseButton(messages.cancel_btn);
        this.appendAcceptButton(acceptLabel);

        if (props && (props.width || props.height)) {
            const style = (this.node.children.item(0) as HTMLDivElement).style;
            style.justifyContent = "flex-start";
            if (props.width) {
                style.maxWidth = `${props.width}px`;
                style.width = `${props.width}px`;
            }

            if (props.height) {
                style.maxHeight = `${props.height}px`;
                style.height = `${props.height}px`;
            }
        }

        this.contentNode.setAttribute(
            "style",
            "justify-content: flex-start; padding-left: 0; padding-right: 0; overflow-y: hidden;"
        );
    }

    protected async init(progress: Progress): Promise<void> {
        if (this.acceptButton)
            this.setEnabled(this.acceptButton, false);
        await this.wizard.init(progress);
        this.wizard.onCompleteChanged(() => this.updateButtons());
        this.wizard.onUpdate(() => this.update());
        this.updateButtons();
        this.update();
    }

    protected doRender(): React.ReactNode {
        this.currentPageDisposable.dispose();
        if (this.wizard.currentPage) {
            this.currentPageDisposable.pushAll([
                this.wizard.currentPage.onCompleteChanged(() => this.updateButtons()),
                this.wizard.currentPage.onUpdate(() => this.update())
            ]);
        }
        return this.wizard.render();
    }

    protected async accept(): Promise<void> {
        if (!this.resolve || this.performingFinish || this.acceptButton?.disabled || !this.wizard.canFinish())
            return;
        try {
            if (this.acceptButton)
                this.setEnabled(this.acceptButton, false);
            this.performingFinish = true;
            this.value = await this.showProgress(undefined, async progress => await this.wizard.performFinish(progress));
        } finally {
            if (this.acceptButton)
                this.setEnabled(this.acceptButton, true);
            this.performingFinish = false;
        }
        if (this.value) {
            this.wizard.dispose();
            return super.accept();
        }
    }

    close(): void {
        this.wizard.dispose();
        super.close();
    }

    protected updateButtons() {
        if (this.isInProgress())
            return;
        if (this.previousBtn) {
            this.setEnabled(
                this.previousBtn,
                this.wizard.getPreviousPage(this.wizard.currentPage) !== undefined
            );
        }

        if (this.nextBtn) {
            this.setEnabled(
                this.nextBtn,
                this.wizard.currentPage !== undefined &&
                this.wizard.currentPage.complete &&
                this.wizard.getNextPage(this.wizard.currentPage) !== undefined
            );
        }
        if (this.acceptButton)
            this.setEnabled(this.acceptButton, this.wizard.canFinish());

        if (this.oldCurrent !== this.wizard.currentPage) {
            this.oldCurrent = this.wizard.currentPage;
            this.update();
        }
    }

    private appendButton(button: HTMLButtonElement): HTMLButtonElement {
        this.controlPanel.appendChild(button);
        button.classList.add("secondary");
        return button;
    }

}
