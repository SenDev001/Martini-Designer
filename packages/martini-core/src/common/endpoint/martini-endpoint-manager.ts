import { JsonRpcServer } from "@theia/core";
import { MartiniEvent } from "../event/martini-event";

export const martiniEndpointManagerPath = "/services/martini/endpoints";

export enum EndpointType {
    DIR_WATCHER = "dir-watcher",
    EMAIL = "email",
    FTP_CLIENT = "ftp-client",
    FTP_SERVER = "ftp-server",
    HTTP_FILTER = "http-filter",
    JABBER = "jabber",
    JMS_LISTENER = "jms-listener",
    KAFKA_LISTENER = "kafka-listener",
    MONGODB_LISTENER = "mongodb-listener",
    URL_ALIAS = "rest-alias",
    RSS = "rss",
    SCHEDULER = "scheduler",
    TRACKER_RESUBMIT = "tracker-resubmit",
    REDIS_LISTENER = "redis-listener"
}

export interface MartiniEndpoint {
    type: EndpointType;
    name: string;
    service: string;
    packageName: string;
    enabled: boolean;
    modifiable: boolean;
    documentType: string;
    track: boolean;
    replicated: boolean;
    status: "STARTED" | "STOPPED";
}

export namespace MartiniEndpoint {
    export function is(object: any): object is MartiniEndpoint {
        return (
            !!object &&
            typeof object === "object" &&
            "type" in object &&
            "name" in object &&
            "service" in object &&
            "packageName" in object &&
            "enabled" in object &&
            "modifiable" in object &&
            "status" in object
        );
    }
    export function getType(type: string): EndpointType | undefined {
        const typeKey: string | undefined = Object.keys(EndpointType).find(
            (key: string) => EndpointType[key as keyof typeof EndpointType] === type
        );
        return typeKey ? EndpointType[typeKey as keyof typeof EndpointType] : undefined;
    }
}

export const MartiniEndpointManager = Symbol("MartiniEndpointManager");

export interface MartiniEndpointManager extends JsonRpcServer<MartiniEndpointManagerClient> {
    getAll(packageName: string): Promise<MartiniEndpoint[]>;

    get(packageName: string, endpointName: string): Promise<MartiniEndpoint | undefined>;

    start(packageName: string, endpointName: string): Promise<void>;

    stop(packageName: string, endpointName: string): Promise<void>;

    delete(packageName: string, endpointName: string): Promise<void>;

    save(packageName: string, endpoint: MartiniEndpoint): Promise<void>;

    test(packageName: string, endpoint: MartiniEndpoint): Promise<void>;
}

export interface EndpointEvent extends MartiniEvent {
    name: string;
    packageName: string;
    event: "STARTED" | "STOPPED" | "DELETED" | "SAVED";
}

export namespace EndpointEvent {
    export function is(event: MartiniEvent): event is EndpointEvent {
        return event.type === "ENDPOINT";
    }
}

export interface MartiniEndpointManagerClient {
    onEvent(event: EndpointEvent): void;
}

export const getDisplayName = (type: EndpointType): string => {
    switch (type) {
        case EndpointType.DIR_WATCHER:
            return "Directory Watcher";
        case EndpointType.EMAIL:
            return "Email";
        case EndpointType.KAFKA_LISTENER:
            return "Kafka Listener";
        case EndpointType.SCHEDULER:
            return "Scheduler";
        case EndpointType.TRACKER_RESUBMIT:
            return "Tracker Resubmit";
        case EndpointType.JABBER:
            return "Jabber";
        case EndpointType.FTP_CLIENT:
            return "FTP Client";
        case EndpointType.FTP_SERVER:
            return "FTP Server";
        case EndpointType.JMS_LISTENER:
            return "JMS Listener";
        case EndpointType.RSS:
            return "RSS";
        case EndpointType.URL_ALIAS:
            return "URL Alias";
        case EndpointType.MONGODB_LISTENER:
            return "MongoDB Listener";
        case EndpointType.REDIS_LISTENER:
            return "Redis Pub/Sub Listener";
        case EndpointType.HTTP_FILTER:
            return "HTTP Filter";
        default:
            return "Unknown";
    }
};

export interface RssEndpoint extends MartiniEndpoint {
    schedule: string;
    onlyNew: boolean;
    rssUrl: string;
}

export interface EmailSettings {
    host: string;
    port: number;
    username: string;
    password: string;
    ssl: boolean;
}

export interface ReplyEmailSettings extends EmailSettings {
    from: string;
}

export interface EmailEndpoint extends MartiniEndpoint, EmailSettings {
    schedule: string;
    replicated: boolean;
    track : boolean;
    protocol: "imap" | "pop3";
    deleteOnReceive: boolean;
    sendOutputAsReply: boolean;
    sendReplyOnError: boolean;
    replyEmailSettings: ReplyEmailSettings;
}
