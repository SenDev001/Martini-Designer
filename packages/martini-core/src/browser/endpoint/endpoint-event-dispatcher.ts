import {Emitter, Event} from "@theia/core";
import {injectable} from "inversify";
import {EndpointEvent, MartiniEndpointManagerClient} from "../../common/endpoint/martini-endpoint-manager";

export const EndpointEventDispatcher = Symbol("EndpointEventDispatcher");

export interface EndpointEventDispatcher {
    onEndpointEvent: Event<EndpointEvent>;
}

@injectable()
export class EndpointEventDispatcherClient
    implements MartiniEndpointManagerClient, EndpointEventDispatcher {
    private readonly onEndpointEventEmitter = new Emitter<EndpointEvent>();
    readonly onEndpointEvent = this.onEndpointEventEmitter.event;

    onEvent(event: EndpointEvent): void {
        this.onEndpointEventEmitter.fire(event);
    }
}
