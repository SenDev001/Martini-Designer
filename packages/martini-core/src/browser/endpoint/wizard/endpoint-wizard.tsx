import { SelectionService } from "@theia/core";
import { OpenerService } from "@theia/core/lib/browser";
import URI from "@theia/core/lib/common/uri";
import { Field, Form, Formik, FormikErrors, FormikProps } from "formik";
import { inject, injectable, postConstruct } from "inversify";
import { flatten, noop } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import * as Yup from "yup";
import { EndpointType, getDisplayName, MartiniEndpoint, MartiniEndpointManager } from "../../../common/endpoint/martini-endpoint-manager";
import { MartiniPackageManager, PartialMartiniPackage } from "../../../common/package/martini-package-manager";
import { UpDownLabel } from "../../components/up-down-label";
import { FormEffect, FormRow, OnFormChange, validateSchema } from "../../form/form";
import { Progress } from "../../progress/progress-service";
import { AbstractWizard, AbstractWizardPage, Wizard } from "../../wizard/wizard";
import { WizardContribution } from "../../wizard/wizard-contribution";
import { EndpointEditor } from "../editor/endpoint-editor";
import { isEndpointSelection } from "../endpoint-contribution";
import { EndpointListTreeNode } from "../endpoint-navigator-contribution";

@injectable()
export class EndpointWizardContribution implements WizardContribution {
    @inject(SelectionService)
    private selectionService: SelectionService;
    @inject("Factory<EndpointWizard>")
    private factory: () => EndpointWizard;

    readonly wizardType = "new";
    readonly id = "martini-endpoint";
    readonly label = messages.endpoint;
    readonly description = messages.create_endpoint;
    readonly iconClass = "martini-icon martini-endpoint-icon";
    readonly keybinding = "ctrlcmd+alt+n e";
    readonly primary = true;
    readonly menuGroup = "11_package";

    async createWizard(..._: any[]): Promise<Wizard> {
        return this.factory();
    }

    isVisible() {
        const selection = this.selectionService.selection;
        return isEndpointSelection(selection) ||
            (selection instanceof Array &&
                (EndpointListTreeNode.is(selection[0]) ||
                    (PartialMartiniPackage.is(selection[0]) && selection[0].name !== "core")));
    }

}

@injectable()
export class EndpointWizardPage extends AbstractWizardPage {
    @inject(SelectionService)
    private readonly selectionService: SelectionService;
    packageNames: string[] = [];
    existingNames: string[] = [];
    private readonly defaultConfig: EndpointConfig = {
        type: EndpointType.DIR_WATCHER,
        name: "MyEndpoint",
        packageName: "",
    };
    config: EndpointConfig = this.defaultConfig;
    private _existingEndpoints: MartiniEndpoint[] = [];

    @postConstruct()
    protected init() {
        const selection = this.selectionService.selection;
        if (isEndpointSelection(selection)) {
            this.defaultConfig.type = selection[0].type;
            this.defaultConfig.packageName = selection[0].packageName;
        }
        else if (selection instanceof Array) {
            if (EndpointListTreeNode.is(selection[0]))
                this.defaultConfig.packageName = selection[0].martiniPackage.name;
            else if (PartialMartiniPackage.is(selection[0]))
                this.defaultConfig.packageName = selection[0].name;
        }
    }

    render(): React.ReactNode {
        return <CreateEndpointForm
            packageNames={this.packageNames}
            existingNames={this.existingNames}
            defaultConfig={this.defaultConfig}
            onValidate={errors => this.handleValidate(errors)}
            onChange={config => this.handleChange(config)}
        />;
    }

    set existingEndpoints(existingEndpoints: MartiniEndpoint[]) {
        this._existingEndpoints = existingEndpoints;
        this.existingNames = this._existingEndpoints
            .filter(e => e.packageName === this.config.packageName)
            .map(e => e.name);
    }

    private handleValidate(errors: FormikErrors<EndpointConfig>) {
        this.complete = Object.keys(errors).length === 0;
    }

    private handleChange(config: EndpointConfig) {
        const oldPackageName = this.config.packageName;
        this.config = config;
        if (oldPackageName !== config.packageName) {
            this.existingNames = this._existingEndpoints
                .filter(e => e.packageName === config.packageName)
                .map(e => e.name);
            this.onUpdateEmitter.fire();
        }
    }
}

export interface EndpointConfig {
    type: EndpointType;
    name: string;
    packageName: string;
}

export interface CreateEndpointFormProps {
    packageNames: string[];
    existingNames: string[];
    defaultConfig?: EndpointConfig;
    onValidate: (errors: FormikErrors<EndpointConfig>) => void;
    onChange: (config: EndpointConfig) => void;
}

const endpointNameRegExp = /^[a-zA-Z0-9\\_-]+$/;

export const CreateEndpointForm: React.FC<CreateEndpointFormProps> = (
    { packageNames, existingNames, defaultConfig, onValidate, onChange }) => {

    const schema = React.useMemo(() => Yup.object().shape<Omit<EndpointConfig, "type" | "packageName">>({
        name: Yup.string()
            .required()
            .max(50)
            .matches(endpointNameRegExp, messages.only_alphanumeric)
            .test("unique-name", messages.endpoint_name_exists, value => !existingNames.includes(value!))
    }), []);
    const validate = (config: EndpointConfig) => validateSchema(schema, config, onValidate);

    const handleKeyDown = (e: React.KeyboardEvent, props: FormikProps<EndpointConfig>) => {
        if (e.key !== "ArrowUp" && e.key !== "ArrowDown")
            return;
        const elements = e.altKey ? packageNames : Object.values(EndpointType);
        let index = elements.indexOf(e.altKey ? props.values.packageName : props.values.type);
        if (e.key === "ArrowUp")
            index = index - 1 < 0 ? elements.length - 1 : index - 1;
        else if (e.key === "ArrowDown")
            index = index + 1 >= elements.length ? 0 : index + 1;
        const selected = elements[index];
        if (e.altKey) {
            props.setValues({
                ...props.values,
                packageName: selected
            });
        }
        else {
            props.setValues({
                ...props.values,
                type: selected as EndpointType
            });
        }
        e.preventDefault();
    };

    return <Formik
        initialValues={defaultConfig}
        initialTouched={{
            type: true,
            name: true,
            packageName: true
        }}
        validate={validate}
        validateOnMount={true}
        onSubmit={noop}
    >
        {props => (
            <Form style={{
                display: "grid",
                gridColumnGap: 10,
                gridTemplateColumns: "max-content 1fr"
            }}>
                <OnFormChange onChange={onChange} />
                <FormEffect deps={[existingNames]} />
                <FormRow name="packageName" label={messages.package}>
                    <Field as="select" name="packageName" className="theia-select">
                        {packageNames.sort().map(name => <option key={name} value={name}>{name}</option>)}
                    </Field>
                </FormRow>
                <FormRow name="type" label={messages.type}>
                    <Field as="select" name="type" className="theia-select">
                        {Object.values(EndpointType).sort().map(type => (
                            <option key={type}
                                value={type}>{getDisplayName(type)}</option>))}
                    </Field>
                </FormRow>
                <FormRow name="name" label={messages.name} gridTemplateColumns="1fr max-content">
                    <Field
                        type="text"
                        name="name"
                        className="theia-input"
                        autoFocus
                        onKeyDown={(e: React.KeyboardEvent) => handleKeyDown(e, props)} />
                    <UpDownLabel
                        toolTip={messages.pressing_updown_changes_package_or_type} />
                </FormRow>
            </Form>)}
    </Formik>;
};

@injectable()
export class EndpointWizard extends AbstractWizard {
    readonly title = messages.create_endpoint_title;

    initialSize = {
        width: 340
    };

    constructor(
        @inject(EndpointWizardPage)
        private readonly page: EndpointWizardPage,
        @inject(MartiniPackageManager)
        private readonly packageManager: MartiniPackageManager,
        @inject(MartiniEndpointManager)
        private readonly endpointManager: MartiniEndpointManager,
        @inject(OpenerService)
        private readonly openerService: OpenerService
    ) {
        super();
        this.pages.push(page);
    }

    async init(progress: Progress): Promise<void> {
        progress.report({
            message: messages.fetching_martini_packages
        });
        const packages = await this.packageManager.getAll();
        progress.report({
            message: messages.fetching_endpoints
        });
        const endpoints = flatten<MartiniEndpoint>(await Promise.all(
            packages.map(pckage => this.endpointManager.getAll(pckage.name))
        ));
        this.page.existingEndpoints = endpoints;
        this.page.packageNames = packages.filter(pckage => pckage.name !== 'core').map(pckage => pckage.name);
        return super.init(progress);
    }

    async performFinish(): Promise<boolean> {
        const uri: URI = new URI(`${EndpointEditor.URI_SCHEME}://${encodeURI(this.page.config.packageName)}/${encodeURI(this.page.config.name)}`
            + `?type=${this.page.config.type}&new=true`);
        const handler = await this.openerService.getOpener(uri);
        handler.open(uri);
        return true;
    }

}
