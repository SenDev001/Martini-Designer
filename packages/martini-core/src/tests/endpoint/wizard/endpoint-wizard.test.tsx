import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import { noop } from "lodash";
import * as React from "react";
import * as Renderer from "react-test-renderer";
import "reflect-metadata";
import { CreateEndpointForm } from "../../../browser/endpoint/wizard/endpoint-wizard";
import { EndpointType } from "../../../common/endpoint/martini-endpoint-manager";

Enzyme.configure({ adapter: new Adapter() });

test("CreateEndpointForm should be rendered", () => {
    const component = Renderer.create(<CreateEndpointForm
        onValidate={noop}
        onChange={noop}
        existingNames={[]}
        packageNames={["test"]}
        defaultConfig={{
            type: EndpointType.DIR_WATCHER,
            packageName: "test",
            name: "test"
        }}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
