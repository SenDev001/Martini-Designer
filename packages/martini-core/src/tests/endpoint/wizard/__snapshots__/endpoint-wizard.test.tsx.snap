// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEndpointForm should be rendered 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
  style={
    Object {
      "display": "grid",
      "gridColumnGap": 10,
      "gridTemplateColumns": "max-content 1fr",
    }
  }
>
  <label
    htmlFor="packageName"
    onContextMenu={[Function]}
    style={
      Object {
        "marginTop": "3px",
        "verticalAlign": "center",
      }
    }
  >
    Package
  </label>
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "1fr",
        "gridTemplateRows": "max-content 1fr",
      }
    }
  >
    <select
      className="theia-select"
      name="packageName"
      onBlur={[Function]}
      onChange={[Function]}
      value="test"
    >
      <option
        value="test"
      >
        test
      </option>
    </select>
    <div
      className="form-error-message-wrapper"
    />
  </div>
  <label
    htmlFor="type"
    onContextMenu={[Function]}
    style={
      Object {
        "marginTop": "3px",
        "verticalAlign": "center",
      }
    }
  >
    Type
  </label>
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "1fr",
        "gridTemplateRows": "max-content 1fr",
      }
    }
  >
    <select
      className="theia-select"
      name="type"
      onBlur={[Function]}
      onChange={[Function]}
      value="dir-watcher"
    >
      <option
        value="dir-watcher"
      >
        Directory Watcher
      </option>
      <option
        value="email"
      >
        Email
      </option>
      <option
        value="ftp-client"
      >
        FTP Client
      </option>
      <option
        value="ftp-server"
      >
        FTP Server
      </option>
      <option
        value="http-filter"
      >
        HTTP Filter
      </option>
      <option
        value="jabber"
      >
        Jabber
      </option>
      <option
        value="jms-listener"
      >
        JMS Listener
      </option>
      <option
        value="kafka-listener"
      >
        Kafka Listener
      </option>
      <option
        value="mongodb-listener"
      >
        MongoDB Listener
      </option>
      <option
        value="redis-listener"
      >
        Redis Pub/Sub Listener
      </option>
      <option
        value="rest-alias"
      >
        URL Alias
      </option>
      <option
        value="rss"
      >
        RSS
      </option>
      <option
        value="scheduler"
      >
        Scheduler
      </option>
      <option
        value="tracker-resubmit"
      >
        Tracker Resubmit
      </option>
    </select>
    <div
      className="form-error-message-wrapper"
    />
  </div>
  <label
    htmlFor="name"
    onContextMenu={[Function]}
    style={
      Object {
        "marginTop": "3px",
        "verticalAlign": "center",
      }
    }
  >
    Name
  </label>
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "1fr max-content",
        "gridTemplateRows": "max-content 1fr",
      }
    }
  >
    <input
      autoFocus={true}
      className="theia-input"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value="test"
    />
    <div
      style={
        Object {
          "fontSize": "16px",
          "marginLeft": "4px",
        }
      }
      title="Pressing Up or Down arrows while editing the name changes the type, if alt is pressed it changes the package."
    >
      â‡…
    </div>
    <div
      className="form-error-message-wrapper"
    />
  </div>
</form>
`;
